Add auto-attached card

FadhlanR opened the pull request:
Ticket: CS-6421

- [x] top-most card in stack(s) is/are shown in Chat Input as "auto-attached"
- [x] "auto-attached" cards show with a dashed border
- [x] "auto-attached" cards show a tooltip on hover: "Topmost card is shared automatically"
- [x] An "X" is shown at right side of pill and can be used to remove the auto-attached card so that it is not attached when the message is sent
- [x] When the top-most card changes (i.e. the user navigates), any current auto-attached cards are removed and the new top-most card(s) is/are attached
- [x] If a prospective auto-attached card is already explicitly attached by the user via the Attach Card button, it does not appear as auto-attached





https://github.com/cardstack/boxel/assets/12637010/f236eaaa-6979-4594-b145-ae200a503ce1

FadhlanR committed:
Add auto-attached card

FadhlanR committed:
fix test

jurgenwerk commented on packages/host/app/components/ai-assistant/card-picker/index.gts:
```diff
@@ -36,39 +108,31 @@ export default class AiAssistantCardPicker extends Component<Signature> {
+              </:trigger>
+
+              <:content>
+                {{#if (eq card.id @autoAttachedCard.id)}}
```
Is this condition needed?

jurgenwerk commented on packages/host/app/components/matrix/room-input.gts:
```diff
@@ -82,8 +95,23 @@ export default class RoomInput extends Component<RoomArgs> {
   private removeCard(card: CardDef) {
-    let cards = this.cardsToAttach?.filter((c) => c.id !== card.id);
-    this.cardsToSend.set(this.args.roomId, cards?.length ? cards : undefined);
+    // If card doesn't exis in `cardsToAttch`,
```
```suggestion
    // If card doesn't exist in `cardsToAttach`,
```

jurgenwerk commented:
1. I noticed a bug in attached cards, nothing happens when I click on X. Not sure if introduced by this PR.

<img width="441" alt="image" src="https://github.com/cardstack/boxel/assets/273660/1846d4aa-55be-4c00-89ed-b8d13d46c9b8">

2. When I remove the autoattached card, and navigate to other cards in the stack, should the new topmost card now be autoattached? I am not seeing this behaviour here but wondering if it should be like that

2. Does "◻︎ Allow access to the cards you can see at the top of your stacks" still make sense now that we are autoattaching them?

IanCal commented:
This should replace the "share the topmost card" functionality, it'll be part of the context, so we should remove that checkbox.

This updates the message type so that auto-attached and attached cards can both coexist -> https://github.com/cardstack/boxel/pull/1045

Potentially we could change the structure here so that they all sit in "attached" cards and then the concept of an open card is just the ID of what the user is looking at? 

Should we block the index card from being attached?

FadhlanR committed:
Fix remove attached card issue

FadhlanR committed:
Remove 'share topmost card' checkbox

FadhlanR committed:
Display auto-attached card if user navigating to another card

FadhlanR committed:
remove unnecasarry condition

FadhlanR commented:
> 1. I noticed a bug in attached cards, nothing happens when I click on X. Not sure if introduced by this PR.
> 
> <img alt="image" width="441" src="https://private-user-images.githubusercontent.com/273660/306964221-1846d4aa-55be-4c00-89ed-b8d13d46c9b8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg2Njg0NjAsIm5iZiI6MTcwODY2ODE2MCwicGF0aCI6Ii8yNzM2NjAvMzA2OTY0MjIxLTE4NDZkNGFhLTU1YmUtNGMwMC04OWVkLWI4ZDEzZDQ2YzliOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyM1QwNjAyNDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jMWIzNWZkYWQwMGM3ZDg4ZTFiMTg3NzlkNjdhMjliYjMwYzQwNzMxNjkzOTk3MWViYmFjNjFmMWQ1NGY2YzVhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.9NUbHozwH2zXJ9_ci7Ng-jxez64R59RDbeZOtg3eWYY">
> 2. When I remove the autoattached card, and navigate to other cards in the stack, should the new topmost card now be autoattached? I am not seeing this behaviour here but wondering if it should be like that
> 3. Does "◻︎ Allow access to the cards you can see at the top of your stacks" still make sense now that we are autoattaching them?

@jurgenwerk Nice catch! I made a few updates.

- I fixed the removing attached card issue.
- I implemented the behavior when user navigating to another card after removing the auto-attached card. Now it must displays auto-attached card if user navigating to different card.
- I removed the "share the topmost card" checkbox.


https://github.com/cardstack/boxel/assets/12637010/c0a8ad56-ed1f-45bc-abdd-3d68d87d37c3

FadhlanR commented:
> This should replace the "share the topmost card" functionality, it'll be part of the context, so we should remove that checkbox.
> 
> This updates the message type so that auto-attached and attached cards can both coexist -> #1045
> 
> Potentially we could change the structure here so that they all sit in "attached" cards and then the concept of an open card is just the ID of what the user is looking at?
> 
> Should we block the index card from being attached?

It make sense to block the index card from being attached, I'll make a change for that.

Yeah I agree, it seems I have to change the structure here because currently, we can only choose between sending attached cards or sending a message with context for open card and it breaks lots of tests. I think I have to wait until your PR get merged, what do you think @IanCal ?

IanCal commented:
@FadhlanR I think my PR should be passing the linting now, I don't mind which way we do this, if it makes sense we could merge mine then I can take over this branch and update it to match the format, then change openCards to just have the ID?

FadhlanR commented:
> @FadhlanR I think my PR should be passing the linting now, I don't mind which way we do this, if it makes sense we could merge mine then I can take over this branch and update it to match the format, then change openCards to just have the ID?

It sounds good, you can do that. Thanks!

IanCal committed:
Merge branch 'main' into cs-6421-auto-attached-card

IanCal committed:
Use card IDs for the open cards, allow multiple possible patches

IanCal committed:
Open cards are shared just by ID

burieberry reviewed (approved):
I updated the tests yesterday. Do you want me to fix the conflict on your branch?

burieberry commented on packages/host/tests/integration/components/operator-mode-test.gts:
```diff
@@ -777,82 +777,6 @@ module('Integration | operator-mode', function (hooks) {
-
-    await waitFor('[data-test-past-sessions-button]');
-    await click('[data-test-past-sessions-button]');
-    await click('[data-test-enter-room="test_a"]');
```
This will still work but since now a new session is created automatically, you can replace these with a single line `let roomId = await openAiAssistant();` and then use `roomId` when making assertions instead of using `test_a`. 

The other change is I put the matrix tests inside a module because it was frustrating to find them.

IanCal replied:
great thanks!

burieberry committed:
Merge remote-tracking branch 'origin/main' into cs-6421-auto-attached-card

IanCal committed:
Remove openCards and replace with openCardIds

IanCal committed:
Merge branch 'cs-6421-auto-attached-card' of github.com:cardstack/boxel into cs-6421-auto-attached-card

IanCal committed:
No create room button straight away, prettier fixes

IanCal committed:
If no cards are specified, don't check for them

IanCal committed:
Lower general timeouts

burieberry committed:
Merge remote-tracking branch 'origin/main' into cs-6421-auto-attached-card

burieberry committed:
update tests to remove autoattached card

burieberry committed:
update sending the prompt

IanCal committed:
Merge branch 'cs-6421-auto-attached-card' of github.com:cardstack/boxel into cs-6421-auto-attached-card

IanCal committed:
Merge branch 'main' into cs-6421-auto-attached-card

habdelra commented:
I noticed that the failing matrix tests are recently added tests. I just double checked and these are passing for me in main. I wonder if the upstream merge worked properly?

burieberry commented:
> I noticed that the failing matrix tests are recently added tests. I just double checked and these are passing for me in main. I wonder if the upstream merge worked properly?

I think I see where it's failing. Since in this PR there's an autoattached card, there are 2 cards instead of 1 card. There's a `removeAutoAttachedCard` helper to get around that. I can update. 

I think ideally we shouldn't be auto-attaching the index card, but I don't know what the decision was about this.

habdelra commented:
> > I noticed that the failing matrix tests are recently added tests. I just double checked and these are passing for me in main. I wonder if the upstream merge worked properly?
> 
> I think I see where it's failing. Since in this PR there's an autoattached card, there are 2 cards instead of 1 card. There's a `removeAutoAttachedCard` helper to get around that. I can update.
> 
> I think ideally we shouldn't be auto-attaching the index card, but I don't know what the decision was about this.

agreed. it's just blindly including the top most card including if the index card is the top most card. i don't think that was an actual decision anyone made. i think it's a bug to be honest.

burieberry committed:
update new tests to remove autoattached card

IanCal commented:
I'll raise a new ticket and pr for ignoring the index card
